### PR TITLE
nfs: add nodeserver

### LINKED ISSUE #3182: nfs: Add nodeserver for nfs node operation within cephcsi
# Describe the feature you'd like to have #

Create a new nfsplugin nodeserver with cephcsi

# What is the value to the end user? (why is it a priority?) #

This will help support the following :
- pod networking
- filesystem encryption?
- probably moving nfs server ip to configmap
- we'll also be reusing a lot of code from within cephcsi for this

# How will we know we have a good solution? (acceptance criteria) #

- cephcsi nfsplugin working similar nfs driver with added benefit as explained above.
 
# Additional context #

- https://github.com/rook/rook/issues/10372
- https://github.com/rook/rook/pull/10411
- https://github.com/kubernetes-csi/csi-driver-nfs/blob/master/pkg/nfs/nodeserver.go

cc @ceph/ceph-csi-contributors

### PR DESCRIPTION
**Pull request by Rakshith-R**
This commit adds nfs nodeserver capable of
mounting nfs volumes, even with pod networking
using NSenter design similar to rbd and cephfs.
NodePublish, NodeUnpublish, NodeGetVolumeStats
and NodeGetCapabilities have been implemented.

Signed-off-by: Rakshith R <rar@redhat.com>

Resolves: #3182

### CODE REVIEW

File: deploy/nfs/kubernetes/csi-nfsplugin.yaml [hunk 1]
```diff
@@ -2,154 +2,134 @@
 apiVersion: apps/v1
 kind: DaemonSet
 metadata:
-  name: csi-nfs-node
+  name: csi-nfsplugin
```
Comment (Madhu-1) [inline deploy/nfs/kubernetes/csi-nfsplugin.yaml:5]: Document the upgrade scenario as we will have new daemonset and users can delete old daemonsets

File: deploy/nfs/kubernetes/csi-nfsplugin.yaml [hunk 2]
```diff
@@ -2,154 +2,134 @@
...
+        - name: driver-registrar
+          # This is necessary only for systems with SELinux, where
+          # non-privileged sidecar containers cannot access unix domain socket
+          # created by privileged CSI driver container.
+          securityContext:
+            privileged: true
+            allowPrivilegeEscalation: true
+          image: registry.k8s.io/sig-storage/csi-node-driver-registrar:v2.5.1
+          args:
+            - "--v=5"
```
Comment (Madhu-1) [inline deploy/nfs/kubernetes/csi-nfsplugin.yaml]: set log level to 1?

File: internal/nfs/driver/driver.go [hunk 3]
```diff
@@ -43,27 +44,48 @@ func (fs *Driver) Run(conf *util.Config) {
...
 		IS: identity.NewIdentityServer(cd),
-		CS: controller.NewControllerServer(cd),
 	}
+
+	switch {
+	case conf.IsNodeServer:
+		topology, err := util.GetTopologyFromDomainLabels(conf.DomainLabels, conf.NodeID, conf.DriverName)
+		if err != nil {
+			log.FatalLogMsg(err.Error())
+		}
```
Comment (Madhu-1) [inline internal/nfs/driver/driver.go]: topology is not supported for nfs just remove it?

File: internal/nfs/driver/driver.go [hunk 4]
```diff
@@ -43,27 +44,48 @@ func (fs *Driver) Run(conf *util.Config) {
...
+		srv.NS = nodeserver.NewNodeServer(cd, conf.Vtype, topology)
+
+	case conf.IsControllerServer:
+		srv.CS = controller.NewControllerServer(cd)
+	case !conf.IsNodeServer && !conf.IsControllerServer:
+		topology, err := util.GetTopologyFromDomainLabels(conf.DomainLabels, conf.NodeID, conf.DriverName)
```
Comment (Madhu-1) [inline internal/nfs/driver/driver.go]: same here

File: internal/nfs/nodeserver/nodeserver.go [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	mount "k8s.io/mount-utils"
+	netutil "k8s.io/utils/net"
+)
+
+const (
+	defaultMountPermission = uint64(0o777)
```
Comment (Madhu-1) [inline internal/nfs/nodeserver/nodeserver.go]: 777 is wider permission not possible to get rid of it? we removed 777 for both cephfs and rbd

File: internal/nfs/nodeserver/nodeserver.go [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+			server = v
+		case paramShare:
+			baseDir = v
+		case paramClusterID:
+			clusterID = v
+		case mountOptionsField:
+			if v != "" {
+				mountOptions = append(mountOptions, v)
+			}
+		case mountPermissionsField:
```
Comment (Madhu-1) [inline internal/nfs/nodeserver/nodeserver.go]: do we have this field set by storageclass or the controller?

File: internal/nfs/nodeserver/nodeserver.go [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+	defaultNodeServer := csicommon.NewDefaultNodeServer(d, t, topology)
+
+	return &NodeServer{
+		DefaultNodeServer: defaultNodeServer,
+		Mounter:           defaultNodeServer.Mounter,
+	}
+}
+
+// NodePublishVolume mount the volume.
+func (ns *NodeServer) NodePublishVolume(
```
Comment (Madhu-1) [inline internal/nfs/nodeserver/nodeserver.go:62]: question:- cant we have one NodeStage and multiple NodePublish?

File: internal/nfs/nodeserver/nodeserver.go [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+
+	log.DefaultLog("nfs: mounting volumeID(%v) source(%s) targetPath(%s) mountflags(%v)",
+		volumeID, source, mountPoint, mountOptions)
+	if netNamespaceFilePath != "" {
+		_, stderr, err = util.ExecuteCommandWithNSEnter(
+			ctx, netNamespaceFilePath, "mount", args[:]...)
+	} else {
+		_, stderr, err = util.ExecCommand(ctx, "mount", args[:]...)
+	}
+	if err != nil {
```
Comment (Madhu-1) [inline internal/nfs/nodeserver/nodeserver.go:239]: Need to have check for stderr also?

File: internal/nfs/nodeserver/nodeserver.go (hunk 5 above)
Comment (Rakshith-R) [reply internal/nfs/nodeserver/nodeserver.go]: I'll use `0o750` here as default too ?
https://github.com/ceph/ceph-csi/blob/8c5563a9bcef229b6cfe54ae9fbfbbbed70089be/internal/util/util.go#L313-L316

File: internal/nfs/nodeserver/nodeserver.go (hunk 6 above)
Comment (Rakshith-R) [reply internal/nfs/nodeserver/nodeserver.go]: it should be passed through storageclass params, I'll add docs for the same.

File: internal/nfs/nodeserver/nodeserver.go (hunk 7 above)
Comment (Rakshith-R) [reply internal/nfs/nodeserver/nodeserver.go:62]: > question:- cant we have one NodeStage and multiple NodePublish?

We can but from what i remember from discussion with @nixpanic , 
the kernel remembers nfs mount refs, so if we have one nodePublish for that volume, another nodePublish will just increment the refs and use the same. 
@nixpanic Can you please confirm this.

So there is no particular advantage of having NodeStage and NodePublish. 
and to keep it similar to previously used k8s csi-nfs-driver.

File: deploy/nfs/kubernetes/csi-nfsplugin.yaml (hunk 1 above)
Comment (Rakshith-R) [reply deploy/nfs/kubernetes/csi-nfsplugin.yaml:5]: https://github.com/ceph/ceph-csi/issues/3271

File: internal/nfs/nodeserver/nodeserver.go [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+		if strings.Contains(err.Error(), "invalid argument") {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mountPermissions > 0 && !req.GetReadonly() {
+		err = chmodIfPermissionMismatch(ctx, targetPath, os.FileMode(mountPermissions))
+
```
Comment (Madhu-1) [inline internal/nfs/nodeserver/nodeserver.go]: new line not required.

File: internal/nfs/nodeserver/nodeserver.go [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+// validateNodePublishVoluemRequest validates node publish volume request.
+func validateNodePublishVoluemRequest(req *csi.NodePublishVolumeRequest) error {
+	switch {
+	case req.GetVolumeId() == "":
+		return fmt.Errorf("volume ID missing in request")
+	case req.GetVolumeCapability() == nil:
+		return fmt.Errorf("volume capability missing in request")
+	case req.GetTargetPath() == "":
+		return fmt.Errorf("target path missing in request")
```
Comment (Madhu-1) [inline internal/nfs/nodeserver/nodeserver.go]: use  errors.New?

**Review by Madhu-1 — COMMENTED** [review]
changes LGTM. let's sort out https://github.com/ceph/ceph-csi/pull/3258/files#r930986164 and get this in.

File: internal/nfs/nodeserver/nodeserver.go (hunk 5 above)
Comment (Rakshith-R) [reply internal/nfs/nodeserver/nodeserver.go]:
```
[1mSTEP[0m: create a PVC and bind it to an app with normal user
Jul 28 11:20:37.978: INFO: Waiting up to &PersistentVolumeClaim{ObjectMeta:{nfs-2540  nfs-2540    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Spec:PersistentVolumeClaimSpec{AccessModes:[ReadWriteMany],Resources:ResourceRequirements{Limits:ResourceList{},Requests:ResourceList{storage: {{1073741824 0} {<nil>} 1Gi BinarySI},},},VolumeName:,Selector:nil,StorageClassName:*csi-nfs-sc,VolumeMode:nil,DataSource:nil,DataSourceRef:nil,},Status:PersistentVolumeClaimStatus{Phase:,AccessModes:[],Capacity:ResourceList{},Conditions:[]PersistentVolumeClaimCondition{},AllocatedResources:ResourceList{},ResizeStatus:nil,},} to be in Bound state
Jul 28 11:20:37.978: INFO: waiting for PVC nfs-2540 (0 seconds elapsed)
Jul 28 11:20:39.983: INFO: waiting for PVC nfs-2540 (2 seconds elapsed)
Jul 28 11:20:39.990: INFO: Waiting for PV pvc-aab980e4-1de5-4e70-882b-acaa1e8ed549 to bind to PVC nfs-2540
Jul 28 11:20:39.990: INFO: Waiting up to timeout=10m0s for PersistentVolumeClaims [nfs-2540] to have phase Bound
Jul 28 11:20:39.992: INFO: PersistentVolumeClaim nfs-2540 found and phase=Bound (2.726822ms)
Jul 28 11:20:39.992: INFO: Waiting up to 10m0s for PersistentVolume pvc-aab980e4-1de5-4e70-882b-acaa1e8ed549 to have phase Bound
Jul 28 11:20:39.995: INFO: PersistentVolume pvc-aab980e4-1de5-4e70-882b-acaa1e8ed549 found and phase=Bound (2.674095ms)
Jul 28 11:20:40.009: INFO: Waiting up to pod-run-as-non-root to be in Running state
Jul 28 11:20:56.020: INFO: ExecWithOptions {Command:[/bin/sh -c echo testing > /target/testing] Namespace:nfs-2540 PodName:pod-run-as-non-root ContainerName:write-pod Stdin:<nil> CaptureStdout:true CaptureStderr:true PreserveWhitespace:true Quiet:false}
Jul 28 11:20:56.020: INFO: >>> kubeConfig: /root/.kube/config
Jul 28 11:20:56.021: INFO: ExecWithOptions: Clientset creation
Jul 28 11:20:56.021: INFO: ExecWithOptions: execute(POST https://192.168.39.123:8443/api/v1/namespaces/nfs-2540/pods/pod-run-as-non-root/exec?command=%2Fbin%2Fsh&command=-c&command=echo+testing+%3E+%2Ftarget%2Ftesting&container=write-pod&container=write-pod&stderr=true&stdout=true)
Jul 28 11:20:56.187: INFO: stdErr occurred: /bin/sh: /target/testing: Permission denied

Jul 28 11:20:56.187: FAIL: failed to validate normal user CephFS pvc and application binding: failed to exec command in pod: command terminated with exit code 1
```

Hey @Madhu-1 , I dont think that will work. 
https://github.com/kubernetes-csi/csi-driver-nfs/blob/c78a98288f3d648cf9d6147fa310444a7f228640/cmd/nfsplugin/main.go#L31
I'll revert to what the csi-nfs-driver used previously 0777. 

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.22/detail/mini-e2e_k8s-1.22/3723/pipeline
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.23/detail/mini-e2e_k8s-1.23/2355/pipeline

Comment (Madhu-1) [reply internal/nfs/nodeserver/nodeserver.go]: Do we have securityContext set in the application pod yaml?

Comment (Madhu-1) [reply internal/nfs/nodeserver/nodeserver.go]: Looks like we have it. lets target this later.

Comment (Rakshith-R) [reply internal/nfs/nodeserver/nodeserver.go]: :+1:

**Comment by Rakshith-R** [comment]
@mergifyio rebase

File: internal/nfs/nodeserver/nodeserver.go (hunk 7 above)
Comment (nixpanic) [reply internal/nfs/nodeserver/nodeserver.go:62]: Indeed, NFS should be able to detect that the NFS-server and export are already mounted somewhere, and the in-kernel NFS-client is re-used (and refcounted).

File: internal/nfs/driver/driver.go [hunk 11]
```diff
@@ -43,27 +44,40 @@ func (fs *Driver) Run(conf *util.Config) {
...
 	server := csicommon.NewNonBlockingGRPCServer()
 	srv := csicommon.Servers{
 		IS: identity.NewIdentityServer(cd),
-		CS: controller.NewControllerServer(cd),
 	}
+
+	switch {
+	case conf.IsNodeServer:
+		srv.NS = nodeserver.NewNodeServer(cd, conf.Vtype)
+
```
Comment (nixpanic) [inline internal/nfs/driver/driver.go]: why an empty line here?

File: internal/nfs/nodeserver/nodeserver.go [hunk 12]
```diff
@@ -0,0 +1,357 @@
...
+	paramClusterID        = "clusterID"
+	mountOptionsField     = "mountOptions"
+	mountPermissionsField = "mountPermissions"
+)
+
+// NodeServer struct of ceph CSI driver with supported methods of CSI
+// node server spec.
+type NodeServer struct {
+	*csicommon.DefaultNodeServer
+	Mounter mount.Interface
```
Comment (nixpanic) [inline internal/nfs/nodeserver/nodeserver.go]: no need to include it if it is already part of the `DefaultNodeServer`?

File: internal/nfs/nodeserver/nodeserver.go [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+// NewNodeServer initialize a node server for ceph CSI driver.
+func NewNodeServer(
+	d *csicommon.CSIDriver,
+	t string,
+) *NodeServer {
+	defaultNodeServer := csicommon.NewDefaultNodeServer(d, t, map[string]string{})
+
+	return &NodeServer{
+		DefaultNodeServer: defaultNodeServer,
+		Mounter:           defaultNodeServer.Mounter,
```
Comment (nixpanic) [inline internal/nfs/nodeserver/nodeserver.go]: I don't think this is needed?

**Review by nixpanic — COMMENTED** [review]
How much of the code was taken from the csi-driver-nfs? It would be nice to reference that, and explain that additional options in the StorageClass come from there.

I'd prefer to have a standard (forked?) driver, and then add the nsenter enhancements as a separate commit. Not sure if it makes sense to spend time on that though...

**Comment by Rakshith-R** [comment]
> How much of the code was taken from the csi-driver-nfs? It would be nice to reference that, and explain that additional options in the StorageClass come from there.
> 
> I'd prefer to have a standard (forked?) driver, and then add the nsenter enhancements as a separate commit. Not sure if it makes sense to spend time on that though...

Most of the code is rather generic CSI vars,func and procedure.
A few lines are inspired from the csi-driver-nfs to keep it backward compatible
for PVC previously mounted using that driver.
I've referenced the csi-driver-nfs in the commit message.

**Comment by Rakshith-R** [comment]
> /retest ci/centos/mini-e2e/k8s-1.22

https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.22/detail/mini-e2e_k8s-1.22/3758/pipeline
failed to start minikube

**Comment by Rakshith-R** [comment]
@mergifyio rebase

File: internal/nfs/nodeserver/nodeserver.go [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+	info, err := os.Lstat(targetPath)
+	if err != nil {
+		return err
+	}
+	perm := info.Mode() & os.ModePerm
+	if perm != mode {
+		log.DebugLog(ctx, "performing chmod targetPath(%s, mode:0%o) with permissions(0%o)",
+			targetPath, info.Mode(), mode)
+		if err = os.Chmod(targetPath, mode); err != nil {
+			return err
```
Comment (Madhu-1) [inline internal/nfs/nodeserver/nodeserver.go]: just return err from here if we do so we can remove else check

**Review by Madhu-1 — DISMISSED** [review]
changes LGTM

File: internal/nfs/nodeserver/nodeserver.go (hunk 14 above)
Comment (Rakshith-R) [reply internal/nfs/nodeserver/nodeserver.go]: done

**Review by Madhu-1 — DISMISSED** [review]

**Comment by Rakshith-R** [comment]
hey @nixpanic , Can you please revisit this pr ? 

The ci is failing due to not getting a node to run on .
https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.22/detail/mini-e2e_k8s-1.22/3792/pipeline

File: internal/nfs/driver/driver.go [hunk 15]
```diff
@@ -43,27 +44,39 @@ func (fs *Driver) Run(conf *util.Config) {
...
 		IS: identity.NewIdentityServer(cd),
-		CS: controller.NewControllerServer(cd),
 	}
+
+	switch {
+	case conf.IsNodeServer:
+		srv.NS = nodeserver.NewNodeServer(cd, conf.Vtype)
+	case conf.IsControllerServer:
+		srv.CS = controller.NewControllerServer(cd)
+	case !conf.IsNodeServer && !conf.IsControllerServer:
```
Comment (nixpanic) [inline internal/nfs/driver/driver.go]: this can be `default:` instead of a case that covers all other possibilities

File: internal/nfs/nodeserver/nodeserver.go [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+	}
+
+	if netutil.IsIPv6String(server) {
+		// if server is IPv6, format to [IPv6].
+		server = fmt.Sprintf("[%s]", server)
+	}
+	source := fmt.Sprintf("%s:%s", server, baseDir)
+
+	err = ns.mountNFS(
+		ctx, volumeID, clusterID, source, targetPath, mountOptions, mountPermissions)
```
Comment (nixpanic) [inline internal/nfs/nodeserver/nodeserver.go]: very weird linebreak, I wonder why formatting tools don't complain.... or our coding style does not follow the tools?

File: internal/nfs/nodeserver/nodeserver.go [hunk 17]
```diff
@@ -0,0 +1,346 @@
...
+	volCap := req.GetVolumeCapability()
+	targetPath := req.GetTargetPath()
+	mountPermissions := defaultMountPermission
+	mountOptions := volCap.GetMount().GetMountFlags()
+	if req.GetReadonly() {
+		mountOptions = append(mountOptions, "ro")
+	}
+
+	var server, baseDir, clusterID string
+	for k, v := range req.GetVolumeContext() {
```
Comment (nixpanic) [inline internal/nfs/nodeserver/nodeserver.go]: some of these parameters are not optional, there needs to be a better verification. Ideally use helper functions for getting the name of the server, the nfs-export (share is Windows/Samba terminology) and others. Return a clear failure when a required option is missing (also easier to add unit-tests for helper functions)

File: internal/nfs/nodeserver/nodeserver.go [hunk 18]
```diff
@@ -0,0 +1,346 @@
...
+		}
+		if strings.Contains(err.Error(), "invalid argument") {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mountPermissions > 0 && !req.GetReadonly() {
+		err = chmodIfPermissionMismatch(ctx, targetPath, os.FileMode(mountPermissions))
```
Comment (nixpanic) [inline internal/nfs/nodeserver/nodeserver.go]: is this required, or can it be done by fsGroupPolicy?

File: internal/nfs/nodeserver/nodeserver.go (hunk 6 above)
Comment (nixpanic) [reply internal/nfs/nodeserver/nodeserver.go]: are docs missing in the csi-driver-nfs that has this option? Please add a PR for it there as well then.

File: internal/nfs/nodeserver/nodeserver.go [hunk 19]
```diff
@@ -0,0 +1,346 @@
...
+func (ns *NodeServer) NodeUnpublishVolume(
+	ctx context.Context,
+	req *csi.NodeUnpublishVolumeRequest,
+) (*csi.NodeUnpublishVolumeResponse, error) {
+	err := util.ValidateNodeUnpublishVolumeRequest(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// Considering kubelet make sure the stage and publish operations
```
Comment (nixpanic) [inline internal/nfs/nodeserver/nodeserver.go]: kubelet is Kubernetes specific. Is this a requirement of the CSI Spec?

File: internal/nfs/nodeserver/nodeserver.go [hunk 20]
```diff
@@ -0,0 +1,346 @@
...
+			targetPath, err)
+	}
+	log.DebugLog(ctx, "nfs: successfully unbounded volume %q from %q",
+		volumeID, targetPath)
+
+	return &csi.NodeUnpublishVolumeResponse{}, nil
+}
+
+// NodeUnstageVolume unstage volume.
+func (ns *NodeServer) NodeUnstageVolume(
```
Comment (nixpanic) [inline internal/nfs/nodeserver/nodeserver.go]: this should not be needed, the DefaultNodeServer should return Unimplemented already.

File: internal/nfs/nodeserver/nodeserver.go [hunk 21]
```diff
@@ -0,0 +1,346 @@
...
+	volumeID, clusterID, source, mountPoint string,
+	mountOptions []string,
+	mountPermissions uint64,
+) error {
+	var (
+		stderr, netNamespaceFilePath string
+		err                          error
+	)
+
+	if clusterID != "" {
```
Comment (nixpanic) [inline internal/nfs/nodeserver/nodeserver.go]: anything that calls this function should probably set `clusterID`, don't fetch additional parameters in a mount function.

File: internal/nfs/nodeserver/nodeserver.go [hunk 22]
```diff
@@ -0,0 +1,346 @@
...
+	if stderr != "" {
+		return fmt.Errorf("nfs: failed to mount %q to %q : stderr %q",
+			source, mountPoint, stderr)
+	}
+
+	return err
+}
+
+// validateNodePublishVoluemRequest validates node publish volume request.
+func validateNodePublishVoluemRequest(req *csi.NodePublishVolumeRequest) error {
```
Comment (nixpanic) [inline internal/nfs/nodeserver/nodeserver.go]: add a unit-test for this

File: internal/nfs/nodeserver/nodeserver.go [hunk 23]
```diff
@@ -0,0 +1,346 @@
...
+		args = append(append(args, "-o"), mountOptions...)
+	}
+
+	log.DefaultLog("nfs: mounting volumeID(%v) source(%s) targetPath(%s) mountflags(%v)",
+		volumeID, source, mountPoint, mountOptions)
+	if netNamespaceFilePath != "" {
+		_, stderr, err = util.ExecuteCommandWithNSEnter(
+			ctx, netNamespaceFilePath, "mount", args[:]...)
+	} else {
+		_, stderr, err = util.ExecCommand(ctx, "mount", args[:]...)
```
Comment (nixpanic) [inline internal/nfs/nodeserver/nodeserver.go]: consider using k8s.io/mount-utils for this, is has lots of error checking and is used elsewhere (RBD) already. Eventually we might be able to extend that package and include the `nsenter` wrapping there too.

File: internal/nfs/nodeserver/nodeserver.go [hunk 24]
```diff
@@ -0,0 +1,346 @@
...
+	case req.GetVolumeContext()[paramServer] == "":
+		return fmt.Errorf("%v missing in request", paramServer)
+	case req.GetVolumeContext()[paramShare] == "":
+		return fmt.Errorf("%v missing in request", paramShare)
+	}
+
+	return nil
+}
+
+// chmodIfPermissionMismatch only performs chmod when permission mismatches.
```
Comment (nixpanic) [inline internal/nfs/nodeserver/nodeserver.go]: This isn't recursive, only for a single directory/file. That probably needs explaining. A unit-test would be good to have too. On the other hand, if fsGroupPolicy or similar can cover this, that would be much better.

File: internal/util/csiconfig.go [hunk 25]
```diff
@@ -95,9 +100,9 @@ func readClusterInfo(pathToConfig, clusterID string) (*ClusterInfo, error) {
 			err, string(content))
 	}
 
-	for _, cluster := range config {
-		if cluster.ClusterID == clusterID {
-			return &cluster, nil
+	for i := range config {
```
Comment (nixpanic) [inline internal/util/csiconfig.go:103]: this looks unrelated to NFS, please remove and create its own PR

File: e2e/nfs.go [hunk 26]
```diff
@@ -43,7 +43,7 @@ var (
 	nfsNodePluginPSP   = "csi-nodeplugin-psp.yaml"
 	nfsRookCephNFS     = "rook-nfs.yaml"
 	nfsDeploymentName  = "csi-nfsplugin-provisioner"
-	nfsDeamonSetName   = "csi-nfs-node"
+	nfsDeamonSetName   = "csi-nfsplugin"
```
Comment (nixpanic) [inline e2e/nfs.go:46]: This can be placed in the same commit as where the name is changed, grouping logical changes together. As a rule, each commit should be fully testable, after each commit e2e should be able to run without failures. We don't have that as a strict requirement, but consider that in future PRs. (if there is breakage, `git bisect ..` can not easily be used anymore)

**Review by nixpanic — CHANGES_REQUESTED** [review]

File: internal/nfs/nodeserver/nodeserver.go (hunk 17 above)
Comment (Rakshith-R) [reply internal/nfs/nodeserver/nodeserver.go]: This is done valiateNodePublishRequest()

File: internal/nfs/nodeserver/nodeserver.go (hunk 22 above)
Comment (nixpanic) [reply internal/nfs/nodeserver/nodeserver.go]: btw, Volume in the name of the function is misspelled...

File: internal/nfs/nodeserver/nodeserver.go (hunk 17 above)
Comment (nixpanic) [reply internal/nfs/nodeserver/nodeserver.go]: oh, right! That feels a little hidden, I prefer to have a check where the value is actually fetched. Multiple helper functions make it simpler, I think. Not a **must**, but would be _nice_ to have the error checking a little more obvious.

File: internal/csi-common/nodeserver-default.go [hunk 27]
```diff
@@ -29,6 +29,7 @@ import (
 
 // DefaultNodeServer stores driver object.
 type DefaultNodeServer struct {
+	csi.UnimplementedNodeServer
```
Comment (Rakshith-R) [inline internal/csi-common/nodeserver-default.go:32]: now the nodestage and unstage calls will be inherited as unimplemented

File: internal/util/csiconfig.go (hunk 25 above)
Comment (Rakshith-R) [reply internal/util/csiconfig.go:103]: creating a different commit, linter seems to be catching this error

**Comment by Rakshith-R** [comment]
Removed mountPersmission and chmod, let's keep this first iteration as simple as possible.

added helpers, testcases and addressed reviews.
PTAL

File: internal/nfs/nodeserver/nodeserver.go [hunk 28]
```diff
@@ -0,0 +1,282 @@
...
+	case req.GetVolumeCapability() == nil:
+		return errors.New("volume capability missing in request")
+	case req.GetTargetPath() == "":
+		return errors.New("target path missing in request")
+	}
+
+	return nil
+}
+
+// getSource validates volume context, extracts and returns sourceq.
```
Comment (nixpanic) [inline internal/nfs/nodeserver/nodeserver.go]: `sourceq`? maybe explain what parameters are expected to be set in the map?

File: internal/nfs/nodeserver/nodeserver_test.go [hunk 29]
```diff
@@ -0,0 +1,163 @@
...
+		volContext map[string]string
+	}
+	tests := []struct {
+		name    string
+		args    args
+		want    string
+		wantErr bool
+	}{
+		{
+			name: "ipv4 address",
```
Comment (nixpanic) [inline internal/nfs/nodeserver/nodeserver_test.go:115]: this does not need to be an IP-address, it can be a hostname too. Please add a test-case for that.

**Review by nixpanic — CHANGES_REQUESTED** [review]
some minor modifications needed, looks pretty complete to me now. Thanks!

File: internal/nfs/nodeserver/nodeserver.go (hunk 28 above)
Comment (Rakshith-R) [reply internal/nfs/nodeserver/nodeserver.go]: done

File: internal/nfs/nodeserver/nodeserver_test.go (hunk 29 above)
Comment (Rakshith-R) [reply internal/nfs/nodeserver/nodeserver_test.go:115]: done

File: internal/nfs/nodeserver/nodeserver.go (hunk 22 above)
Comment (nixpanic) [reply internal/nfs/nodeserver/nodeserver.go]: still a typo in the name of the function `validateNodePublishVoluemRequest`

File: internal/nfs/nodeserver/nodeserver_test.go [hunk 30]
```diff
@@ -0,0 +1,174 @@
...
+		name    string
+		args    args
+		want    string
+		wantErr bool
+	}{
+		{
+			name: "hostname as address",
+			args: args{
+				volContext: map[string]string{
+					paramServer: "hostname.io",
```
Comment (nixpanic) [inline internal/nfs/nodeserver/nodeserver_test.go]: nit: use `example.io` or other official `example.*` hostnames.

**Review by nixpanic — CHANGES_REQUESTED** [review]
_so close!_ just a few nits left.

File: internal/nfs/nodeserver/nodeserver_test.go (hunk 30 above)
Comment (Rakshith-R) [reply internal/nfs/nodeserver/nodeserver_test.go]: done :)

**Review by nixpanic — APPROVED** [review]

**Comment by Rakshith-R** [comment]
@mergifyio rebase 

added a commit to use fsgroupPolicy as File for nfs similar to cephfs, 
ptal

**Review by Madhu-1 — APPROVED** [review]

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Comment by Rakshith-R** [comment]
@mergifyio rebase
